### PR TITLE
Replication performance

### PR DESCRIPTION
**Pull request by cd1989**
Improve FetchImages performance for registry adapters

Issue: https://github.com/goharbor/harbor/issues/7982

### CODE REVIEW

File: src/replication/adapter/native/adapter.go [hunk 1]
```diff
@@ -159,37 +162,75 @@ func (a *Adapter) FetchImages(filters []*model.Filter) ([]*model.Resource, error
...
-		}
-		for _, filter := range filters {
-			if err = filter.DoFilter(&vTags); err != nil {
-				return nil, err
+	rawResources := make([]*model.Resource, len(repositories))
+	var wg = new(sync.WaitGroup)
+	ctx, cancel := context.WithCancel(context.Background())
+	var passportsPool = utils.NewPassportsPool(adp.MaxConcurrency, ctx.Done())
+
+	for i, r := range repositories {
```
Comment (renmaosheng) [inline src/replication/adapter/native/adapter.go]: Could we write a function for this for_loop? it seems we have 3 sections are similar code.

File: src/replication/adapter/aliacr/adapter.go [hunk 2]
```diff
@@ -157,42 +159,81 @@ func (a *adapter) FetchImages(filters []*model.Filter) (resources []*model.Resou
...
+				}
+			} else {
+				rawResources[index] = nil
+			}
+		}(i, r)
+	}
+	wg.Wait()
+
+	err = ctx.Err()
+	cancel()
```
Comment (steven-zou) [inline src/replication/adapter/aliacr/adapter.go]: seems this `cancel` can be deferred too?

File: src/replication/adapter/dockerhub/adapter.go [hunk 3]
```diff
@@ -25,16 +28,16 @@ func init() {
...
-		Credential: registry.Credential,
-		Insecure:   registry.Insecure,
-	})
+	client, err := NewClient(registry)
 	if err != nil {
 		return nil, err
 	}
 
-	dockerRegistryAdapter, err := native.NewAdapter(registry)
+	dockerRegistryAdapter, err := native.NewAdapter(&model.Registry{
```
Comment (steven-zou) [inline src/replication/adapter/dockerhub/adapter.go]: what's the reason for exchanging the passed-in parameters?

File: src/replication/adapter/native/adapter.go (hunk 1 above)
Comment (steven-zou) [reply src/replication/adapter/native/adapter.go]: +1,

Can we move those duplicated codes into a unified runner? Seems most of the code are similar.

Comment (cd1989) [reply src/replication/adapter/native/adapter.go]: OK

File: src/replication/adapter/aliacr/adapter.go [hunk 4]
```diff
@@ -157,42 +159,81 @@ func (a *adapter) FetchImages(filters []*model.Filter) (resources []*model.Resou
...
+						Vtags:  filterTags,
+						Labels: []string{},
 					},
-					Vtags:  filterTags,
-					Labels: []string{},
-				},
-			})
+				}
+			} else {
+				rawResources[index] = nil
```
[the remark this answers is not in the document]
Comment (cd1989) [reply src/replication/adapter/aliacr/adapter.go]: To avoid use lock for the `append` op.

File: src/replication/adapter/aliacr/adapter.go (hunk 2 above)
Comment (cd1989) [reply src/replication/adapter/aliacr/adapter.go]: Yes

File: src/replication/adapter/dockerhub/adapter.go (hunk 3 above)
Comment (cd1989) [reply src/replication/adapter/dockerhub/adapter.go]: DockerHub use too different url for v2 api and others. For example, `hub.docker.io` to list namespace, delete repo, and `registry-1.docker.io` for v2 apis.

**Pull request merged**